application/src/main/java/io/redlink/smarti/webservice/ConversationWebservice.java

```diff
@@ -14,264 +14,511 @@
-import io.swagger.annotations.Api;
-import io.swagger.annotations.ApiOperation;
-import io.swagger.annotations.ApiResponse;
-import io.swagger.annotations.ApiResponses;
+import io.redlink.smarti.webservice.pojo.Projection;
+import io.swagger.annotations.*;
+import org.apache.commons.lang3.ObjectUtils;
+import org.apache.commons.lang3.StringUtils;
 import org.bson.types.ObjectId;
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
 import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.data.domain.Page;
 import org.springframework.http.HttpStatus;
 import org.springframework.http.ResponseEntity;
+import org.springframework.security.access.method.P;
 import org.springframework.util.MimeTypeUtils;
 import org.springframework.web.bind.annotation.*;
+import org.springframework.web.util.UriComponentsBuilder;
 
 import java.io.IOException;
-import java.util.Map.Entry;
-import java.util.Objects;
+import java.net.URI;
+import java.util.List;
 import java.util.Optional;
+import java.util.UUID;
 
 
 /**
  *
  */
+@SuppressWarnings("unused")
 @CrossOrigin
 @RestController
 @RequestMapping(value = "conversation",
         produces = MimeTypeUtils.APPLICATION_JSON_VALUE)
 @Api("conversation")
 public class ConversationWebservice {
 
-    private final Logger log = LoggerFactory.getLogger(getClass());
-    
-    @SuppressWarnings("unused")
-    private enum Vote {
-        up(1),
-        down(-1);
-
-        private final int delta;
+    public static final String API_PARAM_CALLBACK = "URI where to POST the result (triggers async processing)";
+    public static final String API_ASYNC_NOTE = " If a 'callback' is provided, the request will be processed async.";
+    public static final String EDITABLE_CONVERSATION_FIELDS = "meta.status, meta.tags, meta.feedback, context.contextType, context.environmentType, context.domain, context.environment.*";
+    private static final String EDITABLE_MESSAGE_FIELDS = "time, origin, content, private, votes, metadata.*";
 
-        Vote(int delta) {
-            this.delta = delta;
-        }
+    private final AsyncExecutionService asyncExecutionService;
+    private final ConversationService conversationService;
+    private final ClientService clientService;
 
-        public int getDelta() {
-            return delta;
-        }
+    @Autowired
+    public ConversationWebservice(AsyncExecutionService asyncExecutionService, ConversationService conversationService, ClientService clientService) {
+        this.asyncExecutionService = asyncExecutionService;
+        this.conversationService = conversationService;
+        this.clientService = clientService;
     }
 
-    //@Autowired
-    //private StoreService storeService;
+    @ApiOperation(value = "list conversations", response = Page.class)
+    @RequestMapping(method = RequestMethod.GET)
+    public Page<Conversation> listConversations(
+            @RequestParam(value = "clientId", required = false) ObjectId owner,
```

Comment:
I thought the client was determined from the user's client assignment. Why is it a parameter then?